File: backend/services/hkid.py
"""The Hong Kong Identity Card check digit.

WHY THIS IS WORTH HAVING. An HKID carries its own checksum, so a transposed
digit is detectable at the moment someone types it rather than when CR refuses
the filing weeks later. CR treats the two halves as separate XML elements --
NNC1 sends `hkid` and `hkidChkDtg` -- which is why `split_hkid` exists: we
store what the operator typed ("A123456(3)") and take it apart on the way out.

THE ALGORITHM. Right-justify the letter prefix to two characters, padding with
a space. Value a letter as A=10 .. Z=35 and the pad as 36. Weight the eight
characters 9, 8, 7 ... 2, add the check digit (where "A" means 10), and the
total is divisible by 11.

    " A123456" -> 36*9 + 10*8 + 1*7 + 2*6 + 3*5 + 4*4 + 5*3 + 6*2 = 481
    481 + 3 = 484 = 44 * 11, so A123456(3) is valid.

WHAT THIS IS NOT. There is no passport equivalent. A passport's only checksums
are in the machine-readable zone and are computed over the whole MRZ line, not
over the number, so a passport number alone cannot be verified -- it gets
format and length checks and nothing more. Saying otherwise would be a
validator that validates nothing.
"""
import re
from typing import Optional

#: One or two letters, six digits, then the check digit -- optionally in
#: parentheses, optionally spaced. Anything else is not an HKID, including the
#: 18-digit Mainland China resident ID numbers that 29 rows in DEV carry under
#: id_type='hkid'.
_HKID = re.compile(r"^([A-Z]{1,2})(\d{6})\s*\(?([0-9A])\)?$")

_PAD_VALUE = 36     # the space used to right-justify a single-letter prefix
_LETTER_OFFSET = 55  # ord("A") - 10


def _value(char: str) -> int:
    if char == " ":
        return _PAD_VALUE
    if char.isalpha():
        return ord(char) - _LETTER_OFFSET
    return int(char)


def _normalise(number: Optional[str]) -> str:
    return (number or "").strip().upper().replace(" ", "")


def check_digit(prefix_and_digits: str) -> Optional[str]:
    """The check digit for an HKID body such as "A123456", or None.

    Returns a single character: "0"-"9", or "A" for ten.
    """
    body = _normalise(prefix_and_digits)
    match = re.fullmatch(r"([A-Z]{1,2})(\d{6})", body)
    if not match:
        return None

    total = sum(
        _value(char) * weight
        for char, weight in zip(match.group(1).rjust(2), range(9, 7, -1))
    ) + sum(
        int(digit) * weight
        for digit, weight in zip(match.group(2), range(7, 1, -1))
    )
    remainder = (11 - total % 11) % 11
    return "A" if remainder == 10 else str(remainder)


def split_hkid(number: Optional[str]) -> Optional[tuple[str, str]]:
    """("A123456", "3") from "A123456(3)", or None if it is not an HKID.

    Does NOT verify the check digit -- `is_valid_hkid` does that. Splitting a
    stored value and judging it are different questions: a legacy row must
    still be readable even when its check digit is wrong.
    """
    match = _HKID.match(_normalise(number))
    if not match:
        return None
    return f"{match.group(1)}{match.group(2)}", match.group(3)


def is_valid_hkid(number: Optional[str]) -> bool:
    """True when `number` is a well-formed HKID whose check digit agrees."""
    parts = split_hkid(number)
    if parts is None:
        return False
    body, given = parts
    return check_digit(body) == given

File: frontend/src/lib/hkid.js
/**
 * The Hong Kong identity card check digit (Brian's B14).
 *
 * A MIRROR of `backend/services/hkid.py`, not a substitute for it. The API is
 * what actually refuses a bad number — this exists so the operator is told
 * while they are looking at the field, rather than after a save that discards
 * the rest of the form.
 *
 * THE ALGORITHM, and the part everyone gets wrong: the letter prefix is
 * RIGHT-JUSTIFIED into two characters. A single-letter prefix takes a leading
 * space, and that space counts as 36 — not zero, and not nothing. Letters run
 * A=10 to Z=35. Weights are 9 down to 2 across the eight characters, and the
 * whole sum including the check digit must be divisible by 11. A check digit
 * of "A" means 10.
 *
 * Measured against DEV: 452 of the 483 stored HKIDs pass, one fails its check
 * digit (`Z351007(9)`), and 30 do not parse — 29 of those being 18-digit
 * Mainland China ID numbers filed under `id_type = 'hkid'`.
 */

/** CR's cap on `indvPptNo`. There is no check digit to go with it. */
export const PASSPORT_MAX = 25

const PATTERN = /^([A-Z]{1,2})(\d{6})\s*\(?([0-9A])\)?$/
const PREFIX_ONLY = /^([A-Z]{1,2})(\d{6})$/

const SPACE_VALUE = 36
const letterValue = ch => ch.charCodeAt(0) - 'A'.charCodeAt(0) + 10

/**
 * The check digit for a prefix-and-six-digits, or null if that is not what
 * this is. Returns a STRING because "A" is a legal check digit.
 */
export function checkDigit(prefixAndDigits) {
  const match = PREFIX_ONLY.exec(String(prefixAndDigits ?? '').trim().toUpperCase())
  if (!match) return null

  const [, prefix, digits] = match
  // Right-justified to two characters; the pad is worth 36.
  const head = prefix.length === 1
    ? [SPACE_VALUE, letterValue(prefix)]
    : [letterValue(prefix[0]), letterValue(prefix[1])]
  const values = [...head, ...digits.split('').map(Number)]

  let total = 0
  values.forEach((value, i) => { total += value * (9 - i) })

  const remainder = (11 - (total % 11)) % 11
  return remainder === 10 ? 'A' : String(remainder)
}

/** Whether a stored HKID is internally consistent. */
export function isValidHkid(number) {
  const match = PATTERN.exec(String(number ?? '').trim().toUpperCase())
  if (!match) return false
  const [, prefix, digits, check] = match
  return checkDigit(prefix + digits) === check
}

/**
 * What is wrong with this identity number, in words — or null.
 *
 * Only HKID and passport are checked. `china_id` and `other` are left alone
 * on purpose: this portal does not hold a validator for every country's
 * documents, and pretending otherwise would reject correct data.
 */
export function idNumberProblem(idType, number) {
  const value = String(number ?? '').trim()
  // Empty is not this function's business — a required field says so itself,
  // and complaining twice about one blank box helps nobody.
  if (!value) return null

  if (idType === 'hkid' && !isValidHkid(value)) {
    return 'This is not a valid Hong Kong identity card number — the check '
      + 'digit does not match. If it is a Mainland or other identity '
      + 'document, change the document type instead of retyping the number.'
  }
  if (idType === 'passport' && value.length > PASSPORT_MAX) {
    return `${value.length} characters — the Companies Registry accepts `
      + `${PASSPORT_MAX} for a passport number.`
  }
  return null
}
